### PR TITLE
[facebook] Add support for user video playlist

### LINKED ISSUE #10563: Full support for facebook
Hello, is it possible to recover all the videos from a facebook page? It is already possible to do this on Youtube and Dailymotion.

Youtube-dl already suppporte downloading and retrieval of information on a single video.
The videos on facebook link are:
- https://www.facebook.com/USERNAME/videos/
- https://m.facebook.com/MisterVOnline/videos/
- https://touch.facebook.com/MisterVOnline/videos/

thank you in advance

### PR DESCRIPTION
**Pull request by ajj8**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [x] Improvement
- [ ] New extractor
- [x] New feature

---

Fixes #10563

I've implemented Facebook user video playlists. Apologies if it's a bit unclean, I'm still a bit of a novice. It uses the ajax pagelet API for full compatibility (https://www.facebook.com/api/graphql/ didn't support some pages for some reason). A limitation is that you must be signed in (i.e. via --cookies) to extract videos from user profiles instead of a page.

### CODE REVIEW

**Comment by dstftw** [comment]
Read coding conventions.

**Comment by ajj8** [comment]
> Read coding conventions.

Can you at least be more specific? I have tried to follow the coding conventions as much as I can. I would appreciate some feedback since I'm not terribly well versed in programming!

**Comment by ajj8** [comment]
@dstftw Can you help me out here? What needs to be changed to be compliant enough with your coding conventions. I've read through them and tried to comply with them as much as I could, but for instance due to the messy nature of Facebook's JS API it's hard to use regex which would consistently extract a token for the next page, so I'm forced to iterate through a nested dictionary which may not always be the same.

File: youtube_dl/extractor/facebook.py [hunk 1]
```diff
@@ -509,3 +512,116 @@ def _real_extract(self, url):
         return self.url_result(
             compat_urllib_parse_unquote(self._match_id(url)),
             FacebookIE.ie_key())
+
+
+class FacebookUserIE(InfoExtractor):
+    _VALID_URL = r'(?P<url>https?://(?:[^/]+\.)?facebook\.com/(?:pg/)?(?P<id>[^/?#&]+))/videos(?!/[\w\.])'
+    IE_NAME = "facebook:user"
```
Comment (dstftw) [inline youtube_dl/extractor/facebook.py]: Single quotes.

File: youtube_dl/extractor/facebook.py [hunk 2]
```diff
@@ -509,3 +512,116 @@ def _real_extract(self, url):
         return self.url_result(
             compat_urllib_parse_unquote(self._match_id(url)),
             FacebookIE.ie_key())
+
+
+class FacebookUserIE(InfoExtractor):
+    _VALID_URL = r'(?P<url>https?://(?:[^/]+\.)?facebook\.com/(?:pg/)?(?P<id>[^/?#&]+))/videos(?!/[\w\.])'
```
Comment (dstftw) [inline youtube_dl/extractor/facebook.py]: No, override `suitable`.

File: youtube_dl/extractor/facebook.py [hunk 3]
```diff
@@ -24,6 +26,7 @@
     sanitized_Request,
     try_get,
     urlencode_postdata,
+    merge_dicts,
```
Comment (dstftw) [inline youtube_dl/extractor/facebook.py]: Alphabetic.

File: youtube_dl/extractor/facebook.py [hunk 4]
```diff
@@ -509,3 +512,116 @@ def _real_extract(self, url):
...
+        'playlist_mincount': 90,
+    }, {
+        # profile (requires login)
+        'url': 'https://www.facebook.com/zuck/videos',
+        'only_matching': True,
+    }]
+
+    def _real_extract(self, url):
+        mobj = re.match(self._VALID_URL, url)
+        user_id = mobj.group('id')
```
Comment (dstftw) [inline youtube_dl/extractor/facebook.py]: `_match_id`.

File: youtube_dl/extractor/facebook.py [hunk 5]
```diff
@@ -509,3 +512,116 @@ def _real_extract(self, url):
...
+
+        page = self._download_webpage(
+            url, user_id, 'Downloading user webpage')
+        fb_url = self._html_search_meta(
+            'al:android:url', page, default=None)
+        fb_url_mobj = re.match(r'fb://(?P<type>page|profile)/(?P<id>\d+)', fb_url)
+        if not fb_url_mobj:
+            raise ExtractorError('Could not extract page ID', expected=False)
```
Comment (dstftw) [inline youtube_dl/extractor/facebook.py]: `_search_regex`.

File: youtube_dl/extractor/facebook.py [hunk 6]
```diff
@@ -509,3 +512,116 @@ def _real_extract(self, url):
...
+        page_id = fb_url_mobj.group('id')
+        fb_dtsg_ag = self._search_regex(
+            r'"async_get_token":"([\w\-:]+)"',
```
Comment (dstftw) [inline youtube_dl/extractor/facebook.py]: Relax.

File: youtube_dl/extractor/facebook.py [hunk 7]
```diff
@@ -509,3 +512,116 @@ def _real_extract(self, url):
...
+            page, 'fb_dtsg_ag', default=None)
+        pagelet_token = self._search_regex(
+            r'pagelet_token:"([\w\-]+)"',
```
Comment (dstftw) [inline youtube_dl/extractor/facebook.py]: Relax.

File: youtube_dl/extractor/facebook.py [hunk 8]
```diff
@@ -509,3 +512,116 @@ def _real_extract(self, url):
...
+                        merge_dicts(data, {'cursor': cursor}),
+                        separators=(',', ':')),
+                    '__a': 1
+                })
+
+            js_data = self._parse_json(
+                self._search_regex(
+                    r'({.+})', js_data_page,
+                    'js data', default='{}'),
+                user_id, fatal=True)
```
Comment (dstftw) [inline youtube_dl/extractor/facebook.py]: True is default.

File: youtube_dl/extractor/facebook.py [hunk 9]
```diff
@@ -509,3 +512,116 @@ def _real_extract(self, url):
...
+
+            for video in re.findall(
+                    page_re, js_data['payload']):
+                entries.append(
+                    self.url_result(
+                        'https://www.facebook.com%s' % video,
```
Comment (dstftw) [inline youtube_dl/extractor/facebook.py]: `urljoin`.

Comment (dstftw) [inline youtube_dl/extractor/facebook.py]: Filter invalid URLs.

File: youtube_dl/extractor/facebook.py (hunk 2 above)
Comment (dstftw) [inline youtube_dl/extractor/facebook.py]: Do not capture groups you don't use.

File: youtube_dl/extractor/facebook.py [hunk 10]
```diff
@@ -509,3 +512,116 @@ def _real_extract(self, url):
...
+
+        if fb_url_mobj.group('type') == 'page':
+            endpoint = 'PagesVideoHubVideoContainerPagelet'
+            data = {
+                'page': page_id
+            }
+            page_re = r'<td[^>]+>.+?<a href="(?P<url>[^"]+)".+?</td>'
+        elif fb_url_mobj.group('type') == 'profile':
+            if not (fb_dtsg_ag and pagelet_token and collection_token):
+                raise ExtractorError('You must be logged in to extract profile videos', expected=True)
```
Comment (dstftw) [inline youtube_dl/extractor/facebook.py]: You technically can't login with this extractor apart from using cookies.

File: youtube_dl/extractor/facebook.py (hunk 6 above)
Comment (ajj8) [reply youtube_dl/extractor/facebook.py]: How should I relax the regex? The tokens only contain those characters

Comment (ajj8) [reply youtube_dl/extractor/facebook.py]: @dstftw How should I relax the regex? The tokens only contain those characters

File: youtube_dl/extractor/facebook.py (hunk 2 above)
Comment (ajj8) [reply youtube_dl/extractor/facebook.py]: Fixed, I think at least if I understood correctly

File: youtube_dl/extractor/facebook.py (hunk 3 above)
Comment (ajj8) [reply youtube_dl/extractor/facebook.py]: Fixed

File: youtube_dl/extractor/facebook.py (hunk 4 above)
Comment (ajj8) [reply youtube_dl/extractor/facebook.py]: Fixed

File: youtube_dl/extractor/facebook.py (hunk 5 above)
Comment (ajj8) [reply youtube_dl/extractor/facebook.py]: Fixed

File: youtube_dl/extractor/facebook.py (hunk 8 above)
Comment (ajj8) [reply youtube_dl/extractor/facebook.py]: Fixed

File: youtube_dl/extractor/facebook.py (hunk 9 above)
Comment (ajj8) [reply youtube_dl/extractor/facebook.py]: Fixed

Comment (ajj8) [reply youtube_dl/extractor/facebook.py]: Fixed

File: youtube_dl/extractor/facebook.py (hunk 2 above)
Comment (ajj8) [reply youtube_dl/extractor/facebook.py]: Fixed

File: youtube_dl/extractor/facebook.py (hunk 10 above)
Comment (ajj8) [reply youtube_dl/extractor/facebook.py]: I have amended the extractor error to reflect this

File: youtube_dl/extractor/facebook.py (hunk 1 above)
Comment (ajj8) [reply youtube_dl/extractor/facebook.py]: Fixed

**Comment by ajj8** [comment]
Still awaiting review after 3 months. @dstftw @remitamine @phihag

**Comment by ahmedtds** [comment]
Would be nice to get this merged too after latest facebook fixes

**Pull request closed without merging**